Add model instances and allow deduplication

alexykot opened the pull request:
_draft for discussion_

This is another optimisation of the GLTF output - extenting the API to allow deduplicating the meshes.

---
**Purpose**
This PR adds an ability for library users to optimise the resulting GLTF file size by reusing scene assets instead of duplicating them. It extends the `PolyformScene` API, while preserving existing interface and behaviours.

This PR adds `PolyformModelInstance` struct and `[]PolyformModelInstance` slice to the `PolyformScene`. If this slice is not empty - new writer scene creation code path is engaged. On this path - each defined model is added to the scene once as before, but without any nodes defined for any of the meshes, so they are not added to the scene directly.

Instead, each item in the `[]PolyformModelInstance` slice is expected to reference one of the supplied models, and define the placement for that model, including translation, rotation and scaling. Each instance is turned into a node, and nodes are added to the scene. For models that are repeated many times in the scene - this allows to define a model once, and only iterate the placements of the model in the scene, instead of repeating full model mesh data for every placement. 

In cases where scene contains many repeated instances of the same models - this allows to drastically reduce size of the resulting GLTF export.

---
**Notes and questions**
It is technically possible to deduplicate the models by direct comparison same as is already done with materials, but I assume it can be very computationally expensive. Also, I assume that users of the library have control over the scene being built, and have logic in place that can directly iterate the instances of the same mesh, so it will be much easier for the user to do this explicitly it rather than for the library to do it implicitly.

The old behaviour is preserved - if no `Instances` are supplied - each model is assumed to be a unique mesh and added to the scene as is as a single node.

Ideally, I would prefer to change the API slightly from `PolyformScene{ []PolyformModel }` to `PolyformScene{ []*PolyformModel}`. This would allow providing pointers to the same objects in the `Models` and `Instances`. Right now there is a disparity - `Models` are objects, while `Instances` have pointers. Alternatively - I can change `PolyformModelInstance{ Model *PolyformModel }` to `PolyformModelInstance{ ModelIndex int }`, that will be more aligned with the existing structure. I have mild preference for the former over the latter.

This PR tested and works on a simple scene. I need a "30% review" and confirmation of the general approach, then I'll finalise the implementation.

alexykot commented on formats/gltf/writer.go:
```diff
@@ -317,6 +325,142 @@ func rgbToFloatArr(c color.Color) [3]float64 {
+		newNode := Node{Mesh: &meshIndex}
+
+		if instance.Translation != nil {
+			arr := [3]float64{instance.Translation.X(), instance.Translation.Y(), instance.Translation.Z()}
```
For some reason `vector3.Number.ToArr()` returns a slice `[]float64` (in this case), rather than actual Go fixed-length array `[3]float64`. I would prefer to change the `ToArr() []T` signature to `ToArr() [3]T`, that will simpify this bit a little, and be more true to what "to array" implies. But that will be a change of the public API, so not sure how acceptable that would be.

EliCDavis replied:
Yeah, unfortunately, the vector lib isn't in beta, so I'm less comfortable changing signatures

Added `ToFixedArr()` in [v1.8.0](https://github.com/EliCDavis/vector/releases/tag/v1.8.0)

EliCDavis commented:
Woo! Thanks for this! Here are a collection of my thoughts

## Naming

One issue, with some of the naming at least, is that multiple GLTF extensions also use the term "instancing."

* [EXT_mesh_gpu_instancing](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/EXT_mesh_gpu_instancing/README.md) (Complete/Ratified)
* [EXT_instance_features](https://github.com/CesiumGS/glTF/blob/3d-tiles-next/extensions/2.0/Vendor/EXT_instance_features/README.md) (Draft)
  * [Some more on how 3D Tiles uses GLTF](https://github.com/CesiumGS/3d-tiles/blob/main/specification/TileFormats/glTF/README.adoc)
  * [Slides from their presentations at Siggraph](https://www.khronos.org/assets/uploads/developers/presentations/glTF_BOF_-_AEC__Geospatial_incl_gaussian_splatting.pdf)

Some takeaways from `mesh_gpu_instancing`: 

* We don't need to instantiate multiple nodes in the scene to make it work. 
* To define instances, you can include more than just translation, rotation, and scale.
    * > As with vertex attributes, custom instance attributes may be prefixed with an underscore (e.g. _ID, _COLOR, etc.) and used for application-specific effects.
* Maybe we refer to the technique you're presenting as specifically "data instancing"
    * > In authoring tools and data formats, it may refer to reuse of any resource for efficiency, whether that resource represents a GPU draw call or not. We will refer to this technique as data instancing 

## Model API

I think this PR and your previous PR on material optimization is showing some of the cracks in the Model API I tried to put together so developers didn't have to think about Nodes/Samplers/Accessors/Buffers.

`PolyformModel` is more or less a stand-in for the gltf `Node` concept, and I think I made two mistakes when creating it.

1. Not making the Mesh a pointer to begin with.
2. Not including other node properties in the model concept

```go
type PolyformModel struct {
	Name     string
	Mesh     *modeling.Mesh // 1. Make this a pointer
	Material *PolyformMaterial 

	Children     []PolyformModel        // 2. Children concept from Node (Doesn't need to be in this PR)
	Translation  *vector3.Float64       // 2. Concept from Node
	Scale        *vector3.Float64       // 2. Concept from Node
	Quaternion   *quaternion.Quaternion // 2. Concept from Node

	Skeleton   *animation.Skeleton
	Animations []animation.Sequence
}
```

The GLTF spec heavily relies on references between different data types using indices, which emulates the concept of pointers in programming.

## Proposal / Thoughts

We can achieve your proposed "data instancing" optimization without introducing the `PolyformModelInstance` struct. The writer can keep up with what meshes have already been written through pointer references alone, and by including the translation/scale/quaternion

I know this breaks a ton of the examples, I can worry about fixing those.

alexykot commented:
OK, I see what you mean and generally this makes sense - with pointers we can make comparison compute-efficient and not have to push this concern onto the user. Also this will simplify the API and improve ergonomics - at the cost of breaking compatibility. 

And I think using pointers in this way is a sensible API behaviour, i.e. it will be understandable and transparent enough for library users wrt what is happening internally and why. And they keep some control over this behaviour by controlling where those pointers point. This will also allow to access the old behaviour if necessary.

Ok, leave it with me, I'll implement proposed API shape changes in the next few days, and we'll take it from there.

alexykot commented:
Regarding [mesh_gpu_instancing](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/EXT_mesh_gpu_instancing/README.md).

The optimisation metric I'm after right now is the size of the resulting binary.

In my use case - I have a scene that is updated and rerendered repeatedly, e.g. few times a minute, during an extended customer interaction, and most of the interaction is done over the web API. So the customer's browser needs to download and rerender scene updates at every step, and both generation and download are on the hot path. I'm trying to reduce the round trip latency here to minimum, ideally - to subsecond values, although I'm about 10x away from that right now.

So right now I'm working on reducing the binary size to reduce the download time, hence my PRs on size optimisations. 

I have already "integrated" [draco](https://github.com/google/draco) compression into my backend application, but in an extremely clunky way, so the compression itself adds quite a lot of latency, which offsets gains from the binary size reduction. So I'm looking for other ways to improve this, and mesh reuse is an excellent way for me. In my use case I have enormous object duplication (1000x of the same model), and I estimate about 90% binary size reduction with this fully implemented. 

All this is a long way to say that I'm working on a different metric right now, and latency of parsing and rendering the scene on the customer's side after it's downloaded probably will be in my focus somewhere later.

Also, using "instancing" term was a coincidence, I wasn't actually aware of these extensions until you mentioned them 🙂

alexykot commented:
Haha, yes, bolts and screws are the bane of existence. 

Frankly this whole field is new to me, I was working in entirely different area of software world before starting this project, so I just discover things as I explore it. I'll look at 3DTiles, thanks for the advice. Although this will be not right now probably, as I have plenty of other tasks outside of the updates streaming optimisation. It's important, but not the only thing.

Sure, happy to share. My combined model without compression is about 80mb right now, I've dropped it [here](https://drive.google.com/file/d/1p4BkKRMJ54GmAMZWngtmmIUZaYXcQ-er/view?usp=drive_link). If you can suggest any other ways to make it look better and/or load faster - would be great. 

I'll try unwelded cubes, and you might be right. I'm actually using unwelded in places, and those look better than welded ones. I didn't realise this may be because of the normals calculation.

Editing nodes in place would be non-trivial in my case because I have extensive business logic that processes customer requests and produces updates in the logical CAD model, and only after that the logical model is translated into consumable artifacts, visual 3D representation being one of them. I'm not sure translating logical updates into node-level changes would be possible or feasible, but so far I've approximated to this by splitting the model into a list of files. The file I shared is a combined model I use for debugging, while online it's split into ~10 separate parts. This allows to crudely control and reload only what's changed.

There is a limit to what I will be able to achieve now, this project is in pre-MVP stage, so I'm not looking for maximum optimisation yet, only enough to make it look decent on the first release.

EliCDavis commented:
Thanks for the model!

Some observations. (Some you probably already know)

1. 2 million tris isn't great for rendering, but most phones can do that now. I think the performance impact comes from other things.
2. Those screws that the structure "sits" on are 25k tris a piece. Not sure if they need to be that high res. 
3. There's 6000 materials in that model. Is your material optimization not working? This is why it's so slow to render, not the tri count.
4. GLTF isn't the best format given the binary it's base64 encoded, which I think increases the size by like, 30% or something. I would def use GLB.
5. TEXCOORD_0 is not used by 200 of the objects. 
6. Proper GLTF renderers will create an "object" per node. For contexts like Unity, this will end up being 6000 game objects (you can see all this in the hierarchy view in the screenshot), which isn't good for performance.
    1. One way to get around this is to combine all meshes that share the same material. 
        1. This is good because it reduces the number of objects, meshes, and materials being rendered
        2. But it's bad for payload size
    2. The other way would be utilizing the GPU_instancing extension.
        1. Then your limited to viewers that support the extension, but I imagine most do.
7. The `material.pbrMetallicRoughness.baseColorFactor` has a huge number of floating point decimals to describe the RGB colors. For example, `"baseColorFactor": [ 0.8980392156862745, 0.6509803921568628, 0.12549019607843137, 1 ]`. Shrinking this to ``"baseColorFactor": [ 0.898, 0.65, 0.125, 1 ]` shrinks it from 85 characters to 45. Applying this change across the 6000 materials would end up with a reduction of 234kb.
    
![image](https://github.com/user-attachments/assets/9d034212-6b1b-4d2e-be0b-1a012a1fcdbf)

alexykot commented:
Hmm, I've made a mistake when producing this model - I gave you what I think of as "debug" version - a single-file GLTF. 
In production I use split-file GLB, compressed via draco transcoder. My mistake - should have uploaded production build. 
I've uploaded a prod sample [over here](https://drive.google.com/drive/folders/1b44mzl-Z8fXvD4owiiZxMd8vGgxwW5Bh) if you want to have another look. All other things are equal in it, so probably most of what you said still applies.

---
1. Yeah, it worked fine with this count, including on phones, until I added transparent materials (window stand-ins).
2. Yeah, those are ground screws, it's a type of foundation for small structures. Hmm, I've optimised them once already, it was more like 100k a piece in the beginning. But yeah, more should be done there.
3. Yeah, right, I've not bumped my `go.mod` to this commit after it was merged in. Fixed.
4. See above - I made this one as GLTF because I think of it as "debug" format - text based and easy to navigate by hand. Production is GLB+draco. Format flip is a config flag in my system.
5. Hmm, so some objects don't have materials at all. I wonder what those are? I'll have a look, thanks.
6. 
   i. Yes, combining all things with the same material into one mesh is one of the future things I plan to do. But why do you say it's "bad for payload size"? I assume combining many small simple meshes into one large mesh doesn't affect it's overall bytesize when serialised - it's the same number of polygons after all, just rearranged logically. What am I missing here?
    ii. "limited to viewers that support the extension" - It's a hardware limitation, right? What will happen for those who don't support it? Will it fail to render completely or will it just be very slow?
7. Haha, this one is on `polyform`s side :) The API has `color.Color` with simple integer RGB numbers, and it gets `1/x` divided somewhere along the way. I'll see what happens there and fix it, should be an easy fix.

alexykot commented:
OK, I've updated the implementation. Still a draft - no tests and most examples are broken, but it implements the interface changes you've suggested. I've tested on one simple example and it works fine there (on `plumbob`, not included in commit). 

I also like this API more than the my first version, even tho it's breaking compatibility. LMK if this is fine and I'll finish the details.

EliCDavis commented on formats/gltf/writer.go:
```diff
@@ -317,6 +325,112 @@ func rgbToFloatArr(c color.Color) [3]float64 {
 
+func (w *Writer) AddScene(scene PolyformScene) error {
+	// Map to track unique meshes and their indices
+	meshIndices := make(map[*modeling.Mesh]int)
```
I think it's valid to move this to the `Writer` struct. I think it's valid to have multiple scenes reference the same mesh

EliCDavis replied:
And if it is moved to the struct, I think it'd be valid to move this behavior to the `AddMesh` function

EliCDavis commented on formats/gltf/writer.go:
```diff
@@ -317,6 +325,112 @@ func rgbToFloatArr(c color.Color) [3]float64 {
+
+		// Create node with transforms for this model
+		nodeIndex := len(w.nodes)
+		newNode := Node{Mesh: &meshIndex}
```
Oh yeah, we need to include translation, rotation, and scale here

alexykot replied:
It is valid, but pointless?

alexykot commented on formats/gltf/writer.go:
```diff
@@ -317,6 +325,112 @@ func rgbToFloatArr(c color.Color) [3]float64 {
+	// Map to track unique meshes and their indices
+	meshIndices := make(map[*modeling.Mesh]int)
+
+	for _, model := range scene.Models {
```
At this point, before `Children` - does a Model make sense without a Mesh? It doesn't look like it to me, so returning an error if `Mesh == nil`  seems reasonable.

alexykot commented:
OK, I've updated according to your latest comments, but I've discovered an issue that I need your help with.

I've been testing this on `plumbob` example, which I repurposes to compare naive and optimised versions, and I noticed that the optimised version is actually not the same shape as the naive one. 

I'm applying the same transformations manually to create unique meshes for naive version, and then I'm also creating an optimised one with the same mesh and transformations supplied to the model. I expect that the results will be visually identical, but they are not - the shapes are positioned differently, and the rotated mesh is distorted in the optimised version.

Since in the optimised version the transformations are effectively offloaded to the renderer - I've tried two different renderers ([3dviewer](https://3dviewer.net/) and [Babylon](https://sandbox.babylonjs.com/)), and got the same result - the scenes are different when I think they shouldn't be. 

I'm not sure why this is happening, do you have any ideas?

I've committed my testbench `plumbob/main.go`, it produces the two files and in those you can see what I mean.

EliCDavis commented:
Good catch! I had to look at it a while to see the differences. 

One issue is that it's a non-unit quaternion
![image](https://github.com/user-attachments/assets/158ca34a-1693-4dc6-aa14-b47c3cc83b58)

In general, the `FromTheta` function is probably what you want to use.

```go
	rotQuat := quaternion.FromTheta(-math.Pi/2, vector3.New[float64](1, 0, 0))
```

Another Issue:

The GLTF spec is TRS based. Which means Scale is applied first, then rotated, then finally translated. You need to match that in your own editing of the mesh

![image](https://github.com/user-attachments/assets/d18fd337-d4bc-47d7-8cde-8023589f6d54)

With those two changes, you see that the meshes match up!

![image](https://github.com/user-attachments/assets/3022b8e6-3619-49e6-be55-0eb86e4e486e)

If you using VSCode, this is a pretty great extension (it generated the report that told me that you had 200 meshes not using the TEXCOORD)

>Name: glTF Tools
Id: cesium.gltf-vscode
Description: Tools for glTF and GLB 3D models
Version: 2.5.1
Publisher: CesiumJS.org
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=cesium.gltf-vscode

alexykot commented on formats/gltf/writer.go:
```diff
@@ -317,6 +329,134 @@ func rgbToFloatArr(c color.Color) [3]float64 {
+	}
+
+	// Check for empty mesh
+	if model.Mesh.PrimitiveCount() == 0 {
```
There is a test that confirms that empty mesh is correctly rendered into an empty file. 

Shall I preserve that behaviour? Right now in this commit it errors out on empty meshes, but I can revert to previous easily.

EliCDavis replied:
I've def written pipelines in the past where it's valid to generate an empty GLTF (prob where the test came from lol). Let's keep the behavior and not error out.

EliCDavis reviewed (approved):
Woo!
